api/include/opentelemetry/metrics/meter.h

```diff
@@ -0,0 +1,204 @@
+      nostd::string_view name,
+      nostd::string_view description,
+      nostd::string_view unit,
+      const bool enabled,
+      void (*callback)(ObserverResult<T>)) = 0;
+
+  /**
+   * Creates an UpDownSumObserver with the passed characteristics and returns a shared_ptr to
+   * that UpDowNSumObserver.
+   *
+   * @param name the name of the new UpDownSumObserver.
+   * @param description a brief description of what the UpDownSumObserver is used for.
+   * @param unit the unit of metric values following https://unitsofmeasure.org/ucum.html.
+   * @param enabled a boolean value that turns on or off the metric instrument.
+   * @param callback the function to be observed by the instrument.
+   * @return a shared pointer to the created UpDownSumObserver.
+   * @throws NullPointerException if {@code name} is null
+   * @throws IllegalArgumentException if a different metric by the same name exists in this meter.
+   * @throws IllegalArgumentException if the {@code name} does not match spec requirements.
+   */
+  virtual opentelemetry::nostd::shared_ptr<AsynchronousInstrument<T>> NewUpDownSumObserver(
+      nostd::string_view name,
+      nostd::string_view description,
+      nostd::string_view unit,
+      const bool enabled,
+      void (*callback)(ObserverResult<T>)) = 0;
+
+  /**
+   * Creates a ValueObserver with the passed characteristics and returns a shared_ptr to that
+   * ValueObserver.
+   *
+   * @param name the name of the new ValueObserver.
+   * @param description a brief description of what the ValueObserver is used for.
+   * @param unit the unit of metric values following https://unitsofmeasure.org/ucum.html.
+   * @param enabled a boolean value that turns on or off the metric instrument.
+   * @param callback the function to be observed by the instrument.
+   * @return a shared pointer to the created ValueObserver.
+   * @throws NullPointerException if {@code name} is null
+   * @throws IllegalArgumentException if a different metric by the same name exists in this meter.
+   * @throws IllegalArgumentException if the {@code name} does not match spec requirements.
+   */
+  virtual opentelemetry::nostd::shared_ptr<AsynchronousInstrument<T>> NewValueObserver(
+      nostd::string_view name,
+      nostd::string_view description,
+      nostd::string_view unit,
+      const bool enabled,
+      void (*callback)(ObserverResult<T>)) = 0;
+
+  /**
+   * Utility method that allows users to atomically record measurements to a set of
+   * synchronous metric instruments with a common set of labels.
+   *
+   * @param labels the set of labels to associate with this recorder.
+   * @param values a span of pairs where the first element of the pair is a metric instrument
+   * to record to, and the second element is the value to update that instrument with.
+   */
+  virtual void RecordBatch(
+      const trace::KeyValueIterable &labels,
+      const nostd::span<std::pair<nostd::shared_ptr<SynchronousInstrument<T>>,
```

Comment:
It's not ABI safe to have `std::pair` as argument in a virtual method.

Reply:
I think I can get around this instead by taking 3 arguments to the function. The first being a string_view for the labels. Then the second being a span<shared_ptr<SynchronousInstrument<T>>>. And the thirds being a span<T>.
This is essentially having two synchronized spans.

I could overload non-virtual functions to take initializer lists of pairs, then convert that into these two spans and pass them to the original RecordBatch function.

I know this is a bit convoluted but I'm struggling to find another solution here without breaking ABI stability or needing to add to nostd. Any ideas?

Reply:
On second thought. Is it possible to modify the KeyValueIterable class to allow they key to be shared_ptr<SynchronousInstrument<T>>? I could definitely look into doing that as it would be a _much_ cleaner implementation I believe.